### ISSUE TITLE
Matchers using data model utilities based on _get_node() are too slow

### ISSUE THREAD

**Issue by Prabh06** [HazyResearch/fonduer#139]
I was generating candidates from a document where I pretty much needed all the **mentions of the bold text in the first column in the tables:**

lambda function matcher initialisation:
```python
def get_style_attr(html_attr):
    """This function parse a style string and returns a dict of styles.
        eg: 'style=padding-left: 5pt;padding-right: 66pt;font-weight: normal;'
        returns => {
            'padding-left': '5pt',
            'padding-right': '66pt',
            'font-weight': 'normal'
        }
    """
    for attr in html_attr:
        if attr.startswith('style'):
            beg = attr.find('style=') + 6
            return {style.strip().split(':')[0].strip():style.strip().split(':')[1].strip()\
                        for style in attr[beg:].strip().split(';') if len(style.strip()) > 0}

def name_conditions(span):
    if get_parent_tag(span) != 'td':
        return False
    style_attr = get_style_attr(get_attributes(span))
    return style_attr is not None and 'font-weight' in style_attr.keys() and style_attr['font-weight'] == 'bold'

name_lambda_matcher = LambdaFunctionMatcher(func=name_conditions)
```
candidate extractor:

```python
name_label_extractor = CandidateExtractor(name_label, 
                        [name_ngrams, label_ngrams], 
                        [name_lambda_matcher, label_matcher],                              
                        candidate_filter = name_label_filter)

name_label_extractor.apply(docs, parallelism=4)
```
**I am running this on a document with 3620 phrases.**

Is this the expected time for this matcher..? or do you see anything here that is causing this..?

**Comment by lukehsiao** [comment]
> Is this the expected time for this matcher..? or do you see anything here that is causing this..?

This does not seem normal. But, I don't see any obvious errors in what you've written.

Is it possible for you to share a minimal working example with us (e.g. the document and the code up to the candidate extractor?). We may be able to help debug.

**Comment by SenWu** [comment]
Hi @mfboulos ,

Thanks for your information.

> I have at this point simplified my matcher functions to only return true on a single sentence in the Document, and they still run a lot longer than normal.

The underlying strategy to extract mentions is to enumerate all possible phrases in the sentence and verify them using matchers provided by the user. That might be a potential reason why it's slow. But this is a good catch. We will improve the strategy to handle the full sentence case in the next release v0.3.0 (hopefully this week).

> There are a few times that I have interrupted the CandidateExtraction process, it may be related to that somehow.

In v0.2.3, It will clear the corresponding table every time you run the candidate extractor, so yes, it will slow down the process.

> I have also come across the fact that a LOT of memory is being used in this process.

Yes, they are running in parallel as you set in `parallelism=PARALLEL`.

> I am using a candidate model with 5 args, which may be contributing to exponentially increasing the runtime of my program, but at the same time, it seems like what's being saved in the database model is not only the valid candidates that satisfy the matches.

Just to be clear, you are extracting a relation with 5 mentions? If yes, then it will slow down the process since the candidate extractor based do cross product overall mentions and apply throttlers to filter wrong candidates. And also only valid candidates are stored in the database. What did you find there?

Please let us know what you find and some example code will super helpfull as well.

BTW: the new release v0.3.0 will have a huge performance boost.

**Comment by Prabh06** [comment]
@mfboulos a quick tip, we were also facing the issue of candidates exploding if we have more than 2 spans/args, to increase the performance we created 2 separate candidates with 2 spans/args each and after the candidate extraction, we joined them using only 1 span/arg in each

```
eg: 
candidate1("span1", "span2")
candidate2("span3", "span4")

we iterate over the candidates and using helper functions say 
 same_row(span2, span3) --> candidate(span1, span2, span3, span4)
```
This makes sure that cross product happens only among the 2 spans/args

**Comment by Prabh06** [comment]
@lukehsiao @SenWu I have found that adding an if condition to the lambda function matcher increases the time taken exponentially. 
Here is the working code with sample documents. [test.zip](https://github.com/HazyResearch/fonduer/files/2376778/test.zip)

**Comment by lukehsiao** [comment]
Thanks @Prabh06 I've been able to replicate the behavior on my machine. We'll see what we can do.

**Comment by lukehsiao** [comment]
Closed by #152.
